File: bot/exceptions/api_exception.py
"""Exception levée lorsqu'un appel à l'API REST Vicia Home échoue :
erreur réseau, réponse non-2xx, ou corps de réponse invalide.
"""

from __future__ import annotations


class ApiException(Exception):
    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        endpoint: str | None = None,
    ) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.endpoint = endpoint

File: vicia-bot/bot/Core/Exceptions/ApiException.php
<?php

namespace Bot\Core\Exceptions;

/**
 * Class ApiException
 *
 * Levée par Bot\Services\ViciaApiClient lorsqu'un appel à l'API
 * Vicia Home échoue (erreur HTTP 4xx/5xx, réponse malformée,
 * indisponibilité réseau). Conserve le code de statut HTTP d'origine
 * pour permettre aux contrôleurs d'adapter leur réaction (401 → forcer
 * un nouveau lien de compte, 403 → message de droits insuffisants,
 * 5xx → message générique de service indisponible).
 */
class ApiException extends BotException
{
    public function __construct(
        string $userMessage,
        private readonly int $httpStatus,
        string $logMessage = '',
        ?\Throwable $previous = null
    ) {
        parent::__construct($userMessage, $logMessage, $httpStatus, $previous);
    }

    public function httpStatus(): int
    {
        return $this->httpStatus;
    }

    public function isAuthError(): bool
    {
        return $this->httpStatus === 401;
    }

    public function isForbidden(): bool
    {
        return $this->httpStatus === 403;
    }
}
